### PR TITLE
[CI] Enable SYCL End-to-End tests on Windows for post-commit

### PR DESCRIPTION
**Pull request by uditagarwal97**
This PR augments Pavel's work (#7477) on running llvm_test_suite in Windows CI for post-commit testing.
Keeping in mind that the intel/llvm_test_suite repo is getting merged into intel/llvm, this PR tests in-tree SYCL E2E tests.

### CODE REVIEW

File: devops/actions/llvm_test_suite_win/action.yml [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  test_ref:
+    description: 'commit-ish identifier for test repo'
+    required: true
+    default: 'intel'
+  sycl_artifact:
+    description: 'Name of the artifact, that contains compiler toolchain'
+    required: true
+  lit_artifact:
+    description: 'Name of the artifact, that contains LIT tool'
+    required: true
```
Comment (aelovikov-intel) [inline devops/actions/llvm_test_suite_win/action.yml:13]: Do we need that? IIRC, I've switched the Linux action to use the LIT from the repo checkout instead.

File: devops/actions/llvm_test_suite_win/action.yml [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  using: "composite"
+  steps:
+  - name: Checkout LLVM Test Suite
+# TODO: use cached_checkout
+    uses: actions/checkout@v3
+    with:
+      path: llvm_test_suite
+      repository: 'uditagarwal97/llvm-test-suite-intel'
+      ref: 'LitCI'
+#      ref: ${{ inputs.test_ref }}
```
Comment (aelovikov-intel) [inline devops/actions/llvm_test_suite_win/action.yml:35]: Don't need standalone llvm-test-suite support.

File: .github/workflows/sycl_windows_build_and_test.yml [hunk 3]
```diff
@@ -10,11 +10,35 @@ on:
...
+        required: false
+        default: ""
+      lts_ref:
+        type: string
+        required: false
+        default: 'intel'
+      max_parallel:
+        type: number
+        required: false
+        default: 3
```
Comment (aelovikov-intel) [inline .github/workflows/sycl_windows_build_and_test.yml:32]: I think we've recently removed `max_parallel` from Linux workflow.

File: .github/workflows/sycl_post_commit.yml [hunk 4]
```diff
@@ -22,24 +22,21 @@ jobs:
     if: github.repository == 'intel/llvm'
     uses: ./.github/workflows/sycl_gen_test_matrix.yml
     with:
-      lts_config: "l0_gen9"
+      lts_config: "l0_gen9;win_l0_gen12"
       cts_config: "cuda"
   linux_default:
     name: Linux Default
-    if: github.repository == 'intel/llvm'
```
Comment (aelovikov-intel) [inline .github/workflows/sycl_post_commit.yml:29]: Need to move that into a separate commit.

File: .github/workflows/sycl_post_commit.yml [hunk 5]
```diff
@@ -22,24 +22,21 @@ jobs:
...
       build_cache_root: "/__w/llvm"
       build_artifact_suffix: "post_commit"
-      lts_matrix: ${{ needs.test_matrix.outputs.lts_matrix }}
+      lts_matrix: ${{ needs.test_matrix.outputs.lts_lx_matrix }}
       cts_matrix: ${{ needs.test_matrix.outputs.cts_matrix }}
       lts_aws_matrix: ${{ needs.test_matrix.outputs.lts_aws_matrix }}
-      build_configure_extra_args: --hip --cuda --enable-esimd-emulator --cmake-opt="-DSYCL_ENABLE_STACK_PRINTING=ON" --cmake-opt="-DSYCL_LIB_WITH_DEBUG_SYMBOL=ON"
   linux_no_assert:
     name: Linux (no assert)
-    if: github.repository == 'intel/llvm'
```
Comment (aelovikov-intel) [inline .github/workflows/sycl_post_commit.yml:42]: Likewise?

**Review by aelovikov-intel — COMMENTED** [review]
Other than two minor comments, looks good to me.

File: .github/workflows/sycl_gen_test_matrix.yml [hunk 6]
```diff
@@ -34,21 +34,25 @@ on:
         required: false
         default: ${{ github.sha }}
     outputs:
-      lts_matrix:
-        description: "Generated Matrix"
-        value: ${{ jobs.test_matrix.outputs.lts_matrix }}
+      lts_lx_matrix:
+        description: "Generated Linux LTS Matrix"
```
[the remark this answers is not in the document]
Comment (uditagarwal97) [reply .github/workflows/sycl_gen_test_matrix.yml]: Agree, did that!

File: .github/workflows/sycl_post_commit.yml [hunk 7]
```diff
@@ -22,24 +22,21 @@ jobs:
...
     uses: ./.github/workflows/sycl_linux_build_and_test.yml
     secrets: inherit
     with:
       build_cache_root: "/__w/llvm"
       build_artifact_suffix: "post_commit"
-      lts_matrix: ${{ needs.test_matrix.outputs.lts_matrix }}
+      lts_matrix: ${{ needs.test_matrix.outputs.lts_lx_matrix }}
       cts_matrix: ${{ needs.test_matrix.outputs.cts_matrix }}
       lts_aws_matrix: ${{ needs.test_matrix.outputs.lts_aws_matrix }}
-      build_configure_extra_args: --hip --cuda --enable-esimd-emulator --cmake-opt="-DSYCL_ENABLE_STACK_PRINTING=ON" --cmake-opt="-DSYCL_LIB_WITH_DEBUG_SYMBOL=ON"
```
[the remark this answers is not in the document]
Comment (uditagarwal97) [reply .github/workflows/sycl_post_commit.yml:39]: These were some of Pavel's changes that I wasn't sure about. I've removed these unnecessary changes.

File: .github/workflows/sycl_windows_build_and_test.yml [hunk 8]
```diff
@@ -10,11 +10,31 @@ on:
...
         type: string
         required: false
+      lts_matrix:
+        type: string
+        required: false
+        default: "[]"
+      lts_aws_matrix:
+        type: string
+        required: false
+        default: "[]"
```
[the remark this answers is not in the document]
Comment (uditagarwal97) [reply .github/workflows/sycl_windows_build_and_test.yml]: Agree. Fixed!

File: .github/workflows/sycl_windows_build_and_test.yml [hunk 9]
```diff
@@ -113,20 +102,21 @@ jobs:
...
-      if: ${{ always() && !cancelled() && steps.build.outcome == 'success' }}
       uses: actions/upload-artifact@v3
       with:
         name: sycl_windows_default
         path: install
 
   llvm_test_suite:
     needs: build
-    if: ${{ !failure() && inputs.lts_matrix != '[]' }}
+    if: ${{ needs.build.result == 'success' && inputs.lts_matrix != '[]' }}
```
Comment (bader) [inline .github/workflows/sycl_windows_build_and_test.yml:119]:
```suggestion
    if: ${{ inputs.lts_matrix != '[]' }}
```
Don't `need: build` (previous line) guarantee that `needs.build.result == 'success'`?

File: .github/workflows/sycl_windows_build_and_test.yml [hunk 10]
```diff
@@ -50,9 +70,23 @@ jobs:
...
+        cmake --build build --target utils/not/install
+        cmake --build build --target utils/lit/install
+        cmake --build build --target utils/llvm-lit/install
+        cmake --build build --target install-clang-format
+        cmake --build build --target install-clang-tidy
+        cmake --build build --target install-llvm-size
+        cmake --build build --target install-llvm-cov
+        cmake --build build --target install-llvm-profdata
+        cmake --build build --target install-compiler-rt
+        #Copy FileCheck to intall/bin for use by E2E tests.
```
[the remark this answers is not in the document]
Comment (bader) [reply .github/workflows/sycl_windows_build_and_test.yml]: > What does `cmake --build build --target utils/FileCheck/install` (line 77) command do?

I expect `cmake --build build --target utils/FileCheck/install` to copy FileCheck.exe to install directory. Am I missing something?

**Comment by bader** [comment]
Unfortunately, post-commit is broken due to failing LIT test, which blocks E2E tests. We need to resolve post-commit issues to validate the patch.

**Comment by bader** [comment]
> Unfortunately, post-commit is broken due to failing LIT test, which blocks E2E tests. We need to resolve post-commit issues to validate the patch.

I fixed post-commit issue on Windows in https://github.com/intel/llvm/commit/01511a3523a5c0a3d86588067d973bd780cdb371. @uditagarwal97, please, update your branch by merging HEAD of the `sycl` branch.

File: .github/workflows/sycl_windows_build_and_test.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (uditagarwal97) [reply .github/workflows/sycl_windows_build_and_test.yml]: While testing, I was getting several errors related to E2E tests not able to find FileCheck.exe so I added the explicit copy line to resolve the error. I guess `cmake --build build --target utils/FileCheck/install` was having some path-related error while installing. Anyway, I'm now running this command (`cmake --build build --target utils/FileCheck/install`) inside Git's bash (earlier I was using Window's cmd) so **I'll double check if we still need the explicit copy line.**

[the remark this answers is not in the document]
Comment (uditagarwal97) [reply .github/workflows/sycl_windows_build_and_test.yml]: Looks like there's still some problem in `cmake --build build --target utils/FileCheck/install` while installing the binary in the install directory. I tried removing the explicit copy of FileCheck.exe binary but the E2E tests failed with "FileCheck not found error.".
I can try debug this more later, but perhaps, for now, we can get around this problem through the explicit copy of FileCheck binary.

**Review by bader — APPROVED** [review]

File: .github/workflows/sycl_windows_build_and_test.yml [hunk 11]
```diff
@@ -50,9 +70,23 @@ jobs:
...
-        cmake --build build --target sycl-toolchain
-        sccache --show-stats
+        cmake --build build --target deploy-sycl-toolchain
+        cmake --build build --target utils/FileCheck/install
+        cmake --build build --target utils/count/install
+        cmake --build build --target utils/not/install
+        cmake --build build --target utils/lit/install
+        cmake --build build --target utils/llvm-lit/install
+        cmake --build build --target install-clang-format
+        cmake --build build --target install-clang-tidy
```
[the remark this answers is not in the document]
Comment (bader) [reply .github/workflows/sycl_windows_build_and_test.yml]: > I see that I missed the same change in #4911.

Just FYI. I just realized that we need to install clang-format on Linux because `lint` check uses clang-format from our Linux nightly build. There is no use for clang-format on Windows though.

File: .github/workflows/sycl_windows_build_and_test.yml [hunk 12]
```diff
@@ -10,11 +10,26 @@ on:
...
+        required: false
+        default: "[]"
+      lts_cmake_extra_args:
+        type: string
+        required: false
+        default: ""
+      lts_ref:
+        type: string
+        required: false
+        default: 'intel'
```
Comment (aelovikov-intel) [inline .github/workflows/sycl_windows_build_and_test.yml:24]: Not needed.

File: .github/workflows/sycl_windows_build_and_test.yml [hunk 13]
```diff
@@ -80,9 +95,61 @@ jobs:
...
+    - name: Set env
+      run: |
+        git config --system core.longpaths true
+        git config --global core.autocrlf false
+        echo "C:\Program Files\Git\usr\bin" | Out-File -FilePath $env:GITHUB_PATH -Encoding utf8 -Append
+    # TODO: use cached_checkout
+    - uses: actions/checkout@v3
+      with:
+        persist-credentials: false
+        ref: ${{ inputs.build_ref || github.ref }}
```
Comment (aelovikov-intel) [inline .github/workflows/sycl_windows_build_and_test.yml:145]:
```suggestion
        ref: ${{ inputs.build_ref || github.sha }}
```

Linux task was fixed in https://github.com/intel/llvm/pull/8913 .

File: devops/actions/e2e-tests-win/action.yml [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+    shell: cmd
+    run: |
+      mkdir build-e2e
+      set PATH=%GITHUB_WORKSPACE%\install\bin;%PATH%
+      cmake -GNinja -B build-e2e -S.\llvm\sycl\test-e2e -DSYCL_TEST_E2E_TARGETS="${{ inputs.targets }}" -DCMAKE_CXX_COMPILER="clang++" -DLLVM_LIT="..\llvm\llvm\utils\lit\lit.py" ${{ inputs.cmake_args }}
+  - name: Run testing
+    shell: bash
+    run: |
+      # Run E2E tests.
+      export LIT_OPTS="-v --no-progress-bar --show-unsupported --max-time 3600 --time-tests -o results_${{ inputs.results_name_suffix }}.json"
```
Comment (aelovikov-intel) [inline devops/actions/e2e-tests-win/action.yml:41]:
```suggestion
      export LIT_OPTS="-v --no-progress-bar --show-unsupported --max-time 3600 --time-tests"
```

File: devops/actions/e2e-tests-win/action.yml [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+      cmake --build build-e2e --target check-sycl-e2e
+  - name: Upload test results
+    uses: actions/upload-artifact@v1
+    if: always()
+    with:
+      name: lit_results_windows
+      path: build-e2e/results_${{ inputs.results_name_suffix }}.json
```
Comment (aelovikov-intel) [inline devops/actions/e2e-tests-win/action.yml:48]:
```suggestion
```

File: devops/actions/e2e-tests-win/action.yml [hunk 16]
```diff
@@ -0,0 +1,54 @@
+name: 'Run SYCL End-to-End testing'
+description: 'Run SYCL End-to-End testing'
+inputs:
+  sycl_artifact:
+    description: 'Name of the artifact, that contains compiler toolchain'
+    required: true
+  results_name_suffix:
+    description: 'Name suffix of the results artifact'
+    required: true
```
Comment (aelovikov-intel) [inline devops/actions/e2e-tests-win/action.yml:9]:
```suggestion
```

Dropped for Linux in https://github.com/intel/llvm/pull/8822 .

**Review by aelovikov-intel — APPROVED** [review]

File: .github/workflows/sycl_windows_build_and_test.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (bader) [reply .github/workflows/sycl_windows_build_and_test.yml]: Please, add a TODO comment in the code to figure out why building "install" target doesn't work.

**Review by bader — APPROVED** [review]

File: devops/actions/e2e-tests-win/action.yml [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+    run: |
+      mkdir install
+      tar -xf llvm_sycl.tar.gz -C install
+      rm llvm_sycl.tar.gz
+  - name: Configure
+    shell: cmd
+    run: |
+      mkdir build-e2e
+      set PATH=%GITHUB_WORKSPACE%\install\bin;%PATH%
+      cmake -GNinja -B build-e2e -S.\llvm\sycl\test-e2e -DSYCL_TEST_E2E_TARGETS="${{ inputs.targets }}" -DCMAKE_CXX_COMPILER="clang++" -DLLVM_LIT="..\llvm\llvm\utils\lit\lit.py" ${{ inputs.cmake_args }}
```
Comment (aelovikov-intel) [inline devops/actions/e2e-tests-win/action.yml:33]: I'm not sure if we want `clang++` or `clang-cl` here.

Comment (uditagarwal97) [reply devops/actions/e2e-tests-win/action.yml:33]: Previously, I did try experimenting with clang-cl instead of clang++. However, I got a lot of compiler errors. My understanding is that, with Git's Bash, clang++ works fine. However, with U4Win Bash, I had to use `clang-cl` to run the tests. IDK why this happens, but that's something I observed while playing around with the CI.

File: sycl/test-e2e/Regression/reduction_resource_leak_usm.cpp [hunk 18]
```diff
@@ -1,6 +1,6 @@
 // REQUIRES: level_zero, level_zero_dev_kit
-// TODO: ZE_DEBUG=4 produces no output on Windows. Enable when fixed.
 // XFAIL: windows
+// UNSUPPORTED: (ze_debug && windows)
```
Comment (bader) [inline sycl/test-e2e/Regression/reduction_resource_leak_usm.cpp:3]: Could you elaborate more on this change, please?
What is the status of this test on Windows w/ and w/o `ze_debug` config option?

NOTE: I see that ZE_DEBUG is always set on line 6, so I'm not sure if `ze_debug` lit config option used correctly.

Comment (bader) [reply sycl/test-e2e/Regression/reduction_resource_leak_usm.cpp:3]: > Should we disable it on Windows for now and open an internal JIRA for it?

Yes, open a tracker. Please leave original code with the comment and just replace `XFAIL` with `UNSUPPORTED`. 
@bso-intel, FYI.

Comment (uditagarwal97) [reply sycl/test-e2e/Regression/reduction_resource_leak_usm.cpp]: Done!

**Comment by uditagarwal97** [comment]
The failing E2E tests (on Windows) are unrelated and I've filed the bug report in https://github.com/intel/llvm/issues/8934.
@bader should we wait for these tests to get fixed before merging this PR or should we temporarily disable these test cases?

**Comment by bader** [comment]
> The failing E2E tests (on Windows) are unrelated and I've filed the bug report in #8934. @bader should we wait for these tests to get fixed before merging this PR or should we temporarily disable these test cases?

@uditagarwal97, please, disable them. Otherwise, they will fail in post-commits for every commit.

**Comment by uditagarwal97** [comment]
@intel/dpcpp-esimd-reviewers reviews, please. We have disabled some ESIMD E2E test cases as they were failing on Windows (#8934 ).

**Review by sarnex — APPROVED** [review]
esimd lgtm

**Comment by uditagarwal97** [comment]
@intel/llvm-gatekeepers The PR is ready!
